### PR TITLE
fix(store): ease ActionDef signature

### PR DESCRIPTION
**Pull request by deebloo**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
Fix for #794

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #794

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by deebloo** [comment]
@markwhitfeld @splincode are we sure that we want to support object literals for the Action decorator?

**Comment by splincode** [comment]
@deebloo this is very convenient if I don’t want to create a separate action class for this!

```ts
@State(..)
export class CounterState {
  public static increase: string = 'increase';

  @Action({ type: CounterState.increase })
  public increase(ctx) {
    ctx.setState((state) => state++);
  }

}
```

app.component.ts 

```ts
class AppComponent {
  ...  
  ngOnInit() {
    this.store.dispatch({ type: CounterState.increase });
  }
 
}
```

**Comment by markwhitfeld** [comment]
@splincode Does that code sample currently work in plain ngxs?

**Comment by splincode** [comment]
@markwhitfeld Yes, we use it in our projects, because we are uncomfortable creating action classes

**Comment by markwhitfeld** [comment]
Yeah, in my mind it is not a recommended approach but if it is a currently supported then we should continue to support it.
It is not ideal, but as you can see with @splincode it is something that some people prefer.
One of my hopes with v4.0.0 is that we can make the `type` property optional again. I have some ideas how we can acheive this in a production optimised minified code compatible way.

I think it is ok for us to continue to maintain support for the object literal form as an alternative approach in future major versions too.

**Comment by markwhitfeld** [comment]
@deebloo The issue with the fisrt attempt to remove the `type` property is that it relied on the name of the class and this would get scrambled and potentially have duplicates when the code was minified for production.
The basic idea to solve this is that if you pass an action that does not contain a `type` property and it is a class then you can generate a type name and add this type name to the class' prototype object.

The generated name could be based on the class name but then adds some unique identifier (an incrementing number). This would make it useful in dev but also unique in minified prod.

Does this make sense?

**Comment by deebloo** [comment]
Oh yeah. For a brief time we had that. There was a counter used to tag action classes but it was removed. I don't remember the reason though.

**Comment by splincode** [comment]
@markwhitfeld need to fix this before the release

**Comment by eranshmil** [comment]
> Would it be possible for you to find the commit?

I think that's all the relevant PRs:
https://github.com/ngxs/store/pull/212
https://github.com/ngxs/store/pull/211
https://github.com/ngxs/store/pull/210
https://github.com/ngxs/store/pull/209
https://github.com/ngxs/store/pull/208
https://github.com/ngxs/store/pull/207
https://github.com/ngxs/store/pull/206

**Comment by deebloo** [comment]
@splincode ill get to add the tests today

File: packages/store/src/decorators/action.ts [hunk 1]
```diff
@@ -15,10 +18,6 @@ export function Action(actions: ActionDef | ActionDef[], options?: ActionOptions
     for (const action of actions) {
       const type = action.type;
 
-      if (!action.type) {
-        throw new Error(`Action ${action.name} is missing a static "type" property`);
-      }
-
```
Comment (deebloo) [inline packages/store/src/decorators/action.ts]: no need for the runtime check if the compiler will catch the error first

**Pull request merged**